apps/web/pages/api/book/event.ts

```diff
@@ -557,10 +558,25 @@ async function handler(req: NextApiRequest) {
   }
 
   async function createBooking() {
-    // @TODO: check as metadata
-    if (reqBody.web3Details) {
-      const { web3Details } = reqBody;
-      await verifyAccount(web3Details.userSignature, web3Details.userWallet);
+    // Check Eth signature
+    if (eventType.metadata) {
+      const metadata = eventType.metadata as JSONObject;
+      if (metadata.blockchainId && metadata.smartContractAddress) {
+        if (!reqBody.ethSignature) {
+          throw new HttpError({ statusCode: 400, message: "Ethereum signature required." });
+        }
+
+        // @TODO: use the address somewhere in booking creation?
+        const { address, hasBalance } = await verifyEthSig(
+          reqBody.ethSignature,
+          metadata.smartContractAddress as string,
+          metadata.blockchainId as number
+        );
+
+        if (!hasBalance) {
+          throw new HttpError({ statusCode: 400, message: "The wallet doesn't contain enough tokens." });
+        }
+      }
```

Comment:
as discussed earlier, we wanna have all web3 code in the app store and only import it

Reply:
extracted here https://github.com/calcom/cal.com/pull/4019/commits/dd47071d4f5fd693b2dcb77ef7d96c6de3ee4287

Reply:
extracted here https://github.com/calcom/cal.com/commit/dd47071d4f5fd693b2dcb77ef7d96c6de3ee4287